### PR TITLE
Fix crash when `tf.pad` is used with int64 paddings.

### LINKED ISSUE #13506: tf.image.pad_to_bounding_box crashes when passed bounds with dtype int64
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:from pip in virtualenv
- **TensorFlow version (use command below)**:v1.3.0-rc2-20-g0787eee 1.3.0
- **Python version**: '3.5.2 (default, Nov 17 2016, 17:05:23) \n[GCC 5.4.0 20160609]'
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

## Description

Passing arguments of type int64 to `tf.image.pad_to_bounding_box` triggers a crash of the python interpreter. This is a bug because the type required by `tf.image.pad_to_bounding_box` not documented anywhere and just causes a crash with a cryptic error message.

## Sources/Logs

The following snippet crashes the whole python interpreter with a core dump.

    import tensorflow as tf
    i = tf.constant([0, 0, 3, 3], dtype=tf.int64)
    img = tf.ones([1,1,1], dtype=tf.float32)
    sess = tf.Session()
    sess.run(tf.image.pad_to_bounding_box(img, i[0], i[1], i[2], i[3]))

And leaves the following 

    2017-10-05 13:51:24.789715: F tensorflow/core/framework/tensor.cc:493] Check failed: dtype() == expected_dtype (9 vs. 3)

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to fix the issue raised in #13506 where int64 data types for bounds in `tf.image.pad_to_bounding_box` crashes.

The reason of the crash is caused by the fact that int64 was directly converted into int32 without passing through kernel registeration.

This fix fixes the issue by adding `typename Tpadding` to the template and adds appropriate kernels.

This fix fixes #13506.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by frankchn** [comment]
Jenkins, test this please.

**Comment by rryan** [comment]
Happy to review this but I'm on vacation until 10/9.

**Comment by rryan** [comment]
Jenkins, test this please.

File: tensorflow/core/kernels/pad_op.cc [hunk 1]
```diff
@@ -177,11 +193,17 @@ TF_CALL_POD_TYPES(REGISTER_KERNEL);
 namespace functor {
 #define DECLARE_GPU_SPEC(T, Dims)                                         \
   template <>                                                             \
-  void Pad<GPUDevice, T, Dims>::operator()(                               \
+  void Pad<GPUDevice, T, int32, Dims>::operator()(                               \
```
Comment (rryan) [inline tensorflow/core/kernels/pad_op.cc]: please fix the alignment of the \ at the end of the line to match the other lines

File: tensorflow/core/kernels/pad_op.cc [hunk 2]
```diff
@@ -177,11 +193,17 @@ TF_CALL_POD_TYPES(REGISTER_KERNEL);
...
       const GPUDevice& d, typename TTypes<T, Dims>::Tensor output,        \
       typename TTypes<T, Dims>::ConstTensor input,                        \
       Eigen::array<Eigen::IndexPair<int32>, Dims> paddings, T pad_value); \
-  extern template struct Pad<GPUDevice, T, Dims>;
+  extern template struct Pad<GPUDevice, T, int32, Dims>; \
+  template <>                                                             \
+  void Pad<GPUDevice, T, int64, Dims>::operator()(                               \
```
Comment (rryan) [inline tensorflow/core/kernels/pad_op.cc]: please fix the alignment of the \ at the end of the line to match the other lines

**Review by rryan — CHANGES_REQUESTED** [review]
Thanks for the fix! Since this is really a bug in tf.pad, I'm renaming the PR to reflect that.

Could you please add a test to pad_op_test.py that paddings can be int64?
https://github.com/tensorflow/tensorflow/blob/48be6a56d5c49d019ca049f8c48b2df597594343/tensorflow/python/kernel_tests/pad_op_test.py

File: tensorflow/core/kernels/pad_op.cc (hunk 1 above)
Comment (yongtang) [reply tensorflow/core/kernels/pad_op.cc]: Thanks @rryan. The alignment has been fixed.

File: tensorflow/core/kernels/pad_op.cc (hunk 2 above)
Comment (yongtang) [reply tensorflow/core/kernels/pad_op.cc]: Thanks. Done.

**Comment by yongtang** [comment]
Thanks @rryan for the review. The PR has been updated with alignment of `\` fixed. Additional unit tests have also been added in `pad_op_test.py` for `int64`. Without the PR the new test case will trigger the crash.

Please take a look and let me know if there are any issues.

**Review by rryan — APPROVED** [review]

**Comment by rryan** [comment]
LGTM, thanks!

**Comment by rryan** [comment]
Jenkins, test this please.

**Review by benoitsteiner — APPROVED** [review]

**Comment by benoitsteiner** [comment]
Jenkins, test this please.

**Pull request merged**